fix: fix css module definition in typescript declaration file

mcollovati opened the pull request:
## Description

When importing styles in typescript sources, the import should contains the inline parameter (e.g. import styles from './some.css?inline'). The current module definition in types.d.ts does not handle the parameter, so TypeScript errors happen during build.
This change updates the definition in Flow default types.d.ts file and handles upgrade of existing files.

Fixes #18330

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [ ] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

mcollovati commented on flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskGenerateTsDefinitions.java:
```diff
@@ -33,8 +39,42 @@
+
+
+            ************************************************************************
+            *  TypeScript type declaration file 'types.d.ts' have been customized  *
```
fixed

tltv commented:
`customTsDefinition_flowContents_tsDefinitionNotUpdated` fails for me in Windows system. Looks like `content.contains(uncommentedDefaultContent)` condition returns false due to line breaks `\r\n` vs. `\n` line breaks. Screenshot for reference:
![image](https://github.com/vaadin/flow/assets/2182148/6258122e-45d8-4fc1-beaf-d6f3851eff58)


I get following failure and DX when I check out this branch and run `mvn clean install -DskipTests -X`.
Failure is from `test-prod-bundle` module. This was second run. First run failed too to the same error but I'm not sure in which module, it might have been different. 
Second run updates `types.d.ts` by appending new `declare module '*.css?inline' {`  lines. Also it creates a backup of old file. 
Throwing `ExecutionFailedException` even when `types.d.ts` was not customized seems a bit too much.

Smells also like line break issue, but not so sure. I can help testing in Windows.

It works after running it third time.

[INFO] Test Production Bundle ............................. FAILURE [  6.006 s]
[INFO] Flow test Maven plugin with test prod bundle ....... SKIPPED
[INFO] Add-on with frontend files ......................... SKIPPED
[INFO] Add-on with only Java .............................. SKIPPED
[INFO] Reusable custom theme for Express Build tests ...... SKIPPED
[INFO] Reusable custom theme for pre-compiled prod bundle tests SKIPPED
[INFO] Flow Server Production Mode ........................ SKIPPED
[INFO] Flow Bill of Materials ............................. SKIPPED
[INFO] Flow Jandex index .................................. SKIPPED
[INFO] vaadin-spring ...................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:30 min
[INFO] Finished at: 2023-12-27T11:00:52+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vaadin:flow-maven-plugin:24.4-SNAPSHOT:build-frontend (default) on project test-prod-bundle: Could not execute build-frontend goal:
[ERROR]
[ERROR] ***************************************************************************
[ERROR] *  TypeScript type declaration file 'types.d.ts' has been updated to the  *
[ERROR] *  latest version by Vaadin. Previous content has been backed up on       *
[ERROR] *  'types.d.ts.flowBackup' file. Please verify that the updated           *
[ERROR] *  'types.d.ts' file contains configuration needed for your project, and  *
[ERROR] *  then delete the backup file.                                           *
[ERROR] ***************************************************************************
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.vaadin:flow-maven-plugin:24.4-SNAPSHOT:build-frontend (default) on project test-prod-bundle: Could not execute build-frontend goal
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke (Method.java:578)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoFailureException: Could not execute build-frontend goal
    at com.vaadin.flow.plugin.maven.BuildFrontendMojo.execute (BuildFrontendMojo.java:131)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke (Method.java:578)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: com.vaadin.flow.server.ExecutionFailedException:

***************************************************************************
*  TypeScript type declaration file 'types.d.ts' has been updated to the  *
*  latest version by Vaadin. Previous content has been backed up on       *
*  'types.d.ts.flowBackup' file. Please verify that the updated           *
*  'types.d.ts' file contains configuration needed for your project, and  *
*  then delete the backup file.                                           *
***************************************************************************



    at com.vaadin.flow.server.frontend.TaskGenerateTsDefinitions.updateIfContentMissing (TaskGenerateTsDefinitions.java:170)
    at com.vaadin.flow.server.frontend.TaskGenerateTsDefinitions.execute (TaskGenerateTsDefinitions.java:94)
    at com.vaadin.flow.server.frontend.NodeTasks.execute (NodeTasks.java:342)
    at com.vaadin.flow.plugin.base.BuildFrontendUtil.runNodeUpdater (BuildFrontendUtil.java:332)
    at com.vaadin.flow.plugin.maven.BuildFrontendMojo.execute (BuildFrontendMojo.java:129)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke (Method.java:578)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :test-prod-bundle
```

mcollovati commented:
I changed the comparison to normalize EOLs and also added a test for this.
@tltv could you please verify it now works on Windows?

> Throwing ExecutionFailedException even when types.d.ts was not customized seems a bit too much.

This should not happen. If the `types.d.ts` file has the same content as the previous Flow version, it is silently overwritten. It should now be fixed by comparing content ignoring EOL

tltv commented:
> Could you please verify it now works on Windows?

Yes, tests are green now. 

After reverting back the old `types.d.ts` in `test-prod-bundle` module, I still managed to get that same error. However, second and following retries works, so I suspect that I changed the expected code snippet accidentally on the first try.

So it's ok now. It's a little bit unforgiving, as single extra whitespace in wrong place will fail the whole build. On the otherhand, I don't have ideas how to improve and still keep it simple.
